onnxruntime/core/providers/rocm/rocm_profiler.cc

```diff
@@ -0,0 +1,239 @@
+                                            {"grid_y", std::to_string(item.gridY)},
+                                            {"grid_z", std::to_string(item.gridZ)},
+                                            {"block_x", std::to_string(item.workgroupX)},
+                                            {"block_y", std::to_string(item.workgroupY)},
+                                            {"block_z", std::to_string(item.workgroupZ)},
+                                            };
+    addEventRecord(item, profiling_start, args, event_map);
+    kernelLaunches[item.id] = &item;
+  }
+
+  // Async Ops - e.g. "Kernel"
+
+  for (auto& buffer : *d->gpuTraceBuffers_) {
+    const roctracer_record_t* record = (const roctracer_record_t*)(buffer.data_);
+    const roctracer_record_t* end_record = (const roctracer_record_t*)(buffer.data_ + buffer.validSize_);
+    while (record < end_record) {
+      std::unordered_map<std::string, std::string> args;
+      std::string name = roctracer_op_string(record->domain, record->op, record->kind);
+
+      // Add kernel args if we have them
+      auto kit = kernelLaunches.find(record->correlation_id);
+      if (kit != kernelLaunches.end()) {
+        auto &item = *(*kit).second;
+        snprintf(buff, BSIZE, "%p", item.stream);
+        args["stream"] = std::string(buff);
+        args["grid_x"] = std::to_string(item.gridX); 
+        args["grid_y"] = std::to_string(item.gridY); 
+        args["grid_z"] = std::to_string(item.gridZ); 
+        args["block_x"] = std::to_string(item.workgroupX); 
+        args["block_y"] = std::to_string(item.workgroupY); 
+        args["block_z"] = std::to_string(item.workgroupZ); 
+        if (item.functionAddr != nullptr) {
+          name = demangle(hipKernelNameRefByPtr(item.functionAddr, item.stream)).c_str();
+        }
+        else if (item.function != nullptr) {
+          name = demangle(hipKernelNameRef(item.function)).c_str();
+        }
+      }
+
+      // Add copy args if we have them
+      auto cit = copyLaunches.find(record->correlation_id);
+      if (cit != copyLaunches.end()) {
+        auto &item = *(*cit).second;
+        snprintf(buff, BSIZE, "%p", item.stream);
+        args["stream"] = std::string(buff);
+        snprintf(buff, BSIZE, "%p", item.src);
+        args["dst"] = std::string(buff);
+        snprintf(buff, BSIZE, "%p", item.dst);
+        args["src"] = std::string(buff);
+        args["size"] = std::to_string(item.size);
+        args["kind"] = std::to_string(item.kind);
+      }
+
+      EventRecord event{
+          onnxruntime::profiling::KERNEL_EVENT,
+          static_cast<int>(record->device_id),
+          static_cast<int>(record->queue_id),
+          name,
+          static_cast<int64_t>((record->begin_ns - profiling_start) / 1000),
+          static_cast<int64_t>((record->end_ns - record->begin_ns) / 1000),
+          std::move(args)};
+
+       // FIXME: deal with missing ext correlation
```

Comment:
Can you please elaborate a bit the issue here?

Reply:
This is me just mirroring a potential bug I found in cuda_profiler.cc. :)

It looks like there is an implicit assumption that no cuda api call or gpu op can happen outside of an ort span (event category "Session" or "Node").  So every cupti or roctracer correlation ID will have an external id (in that map).  This assumption is made at https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cuda/cuda_profiler.cc#L130

So yeah, I just plagiarized that.  If the external id is missing we should get a default constructed uint64, so 0.  Then there would be cuda/rocm events attributed to external span 0, which may not exist at all or may exist, be legitimate,  and have a lot of unexpected things in it.

Potential issue for a bug hunt.  Hence a comment.

Reply:
Not sure if there is a use case/need for profiling GPU op happen outside of ort span; if not, this may not need to be fixed. How about just adding a comment about the assumption made here?